service/history/taskPriorityAssigner.go

```diff
@@ -0,0 +1,164 @@
+// Copyright (c) 2020 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package history
+
+import (
+	"sync"
+
+	workflow "github.com/uber/cadence/.gen/go/shared"
+	"github.com/uber/cadence/common/cache"
+	"github.com/uber/cadence/common/log"
+	"github.com/uber/cadence/common/log/tag"
+	"github.com/uber/cadence/common/metrics"
+	"github.com/uber/cadence/common/quotas"
+)
+
+type (
+	taskPriorityAssigner interface {
+		Assign(queueTask) error
+	}
+
+	taskPriorityAssignerImpl struct {
+		sync.RWMutex
+
+		currentClusterName string
+		domainCache        cache.DomainCache
+		config             *Config
+		logger             log.Logger
+		scope              metrics.Scope
+		rateLimiters       map[string]quotas.Limiter
+	}
+)
+
+var _ taskPriorityAssigner = (*taskPriorityAssignerImpl)(nil)
+
+const (
+	taskPriorityActiveTransfer int = iota
+	taskPriorityActiveTimer
```

Comment:
Thought about this some more yesterday. I believe what you need here is some sort of priority_class and map the incoming tasks to one of those. Right now, based on your declaration, the task_type and priority are coupled with each other which gives you less flexibility. My proposal is something along the following lines:

- You define some fixed priority_classes and may be also sub-classes. For example, let's say there is a priority_class and a priority_subclass. You can allocate 3 bits for class and another 3 bits for sub-class. To start with, something simple like 3 priority classes - high, default, low would suffice. We can add more later if needed. As for the necessity for sub-classes, it allows you to differentiate between two tasks within same priority class. 

- You assign tasks to priority_class/sub_class i.e. one single integer priority value. For example, both active_transfer_tasks and active_timer_tasks can be high priority but workflow_timeout task can be low_priority. As another example, some domains may be latency sensitive, so, the active_transfer_tasks for these domain can be a high priority sub-class while other active_transfer_tasks can be a low-priority sub-class.

Reply:
Makes sense. Let's sync offline regarding the implementation details.

Reply:
Synced offline. Now using 3 bits for priority_class and another 3 for priority_subclass. For now each level only has three values: high, default and low. We can add more later if necessary.

The mapping from task -> priority is: 
active timer/transfer -> high (priority_class), default (priority_subclass)
throttled (based on domain) timer/transfer -> default (priority_class), default (priority_subclass)
replication, standby timer/transfer -> low (priority_class), default (priority_subclass)

We can change the mapping later. For example, use different subclass based on the exact type of timer/transfer task.